src/library/scala/collection/immutable/ChampHashMap.scala

```diff
@@ -102,8 +106,288 @@ final class HashMap[K, +V] private[immutable] (private val rootNode: MapNode[K,
   }
 
   override def concat[V1 >: V](that: scala.IterableOnce[(K, V1)]): HashMap[K, V1] = {
-    // TODO PERF We could avoid recomputing entry hash's when `that` is another `HashMap`
-    super.concat(that)
+
+    /** Fall back to slow concatenation in case of error, or non-hashmap `that` */
+    def slowConcat = {
+      val builder = mapFactory.newBuilder[K, V1]
+      builder ++= this
+      builder ++= that
+      builder.result()
+    }
+
+    that match {
+      case hm: HashMap[K, V1] =>
+        var newHash = cachedJavaKeySetHashCode + hm.cachedJavaKeySetHashCode
```

Comment:
Add a comment here to explain how this ends up at the right value. e.g. 

"We start with the assumption that the maps are distinct, and then subtract the hash code of keys in  `this` are overwritten by keys in `that`."